agent/tlsfwd/tlsfwd.go

```diff
@@ -0,0 +1,201 @@
+	cancelled atomic.Bool
+	err       atomic.Value
+}
+
+// NewTLSFwd constructs and returns a new TLS forwarder from arguments
+func NewTLSFwd(ports []int, options ...func(*Server) error) (*Server, error) {
+	ctx1, cancel := context.WithCancel(context.Background())
+
+	pts := make([]string, 0)
+	for _, p := range ports {
+		port := ":" + strconv.Itoa(p)
+		pts = append(pts, port)
+	}
+
+	srv := Server{
+		ps:     pts,
+		ctx:    ctx1,
+		cancel: cancel,
+	}
+
+	for _, opt := range options {
+		err := opt(&srv)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// SetRedir will allow http -> https redirection from port 80 to 443
+func SetRedir(s *Server) error {
+	s.redir = true
+	return nil
+}
+
+func getDefaultTLSHandler(c *tls.ClientHelloInfo) sni.Handler {
+	return func(ctx context.Context,
+		conn net.Conn,
+	) error {
+		return tlsHandler(ctx, c, conn)
+	}
+}
+
+func tlsHandler(ctx context.Context, c *tls.ClientHelloInfo,
+	conn net.Conn,
+) error {
+	defer conn.Close()
+
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+	}
+
+	upstream, err := addrFromName(c.ServerName)
+	if err != nil {
+		return err
+	}
+
+	errm := proxy.Forward(ctx, conn, upstream)
+	fmt.Println(errm)
+	return errm
+}
+
+func addrFromName(s string) (netip.AddrPort, error) {
+	upIPs, err := net.LookupIP(s)
+	if err != nil {
+		fmt.Println(err)
+		return netip.AddrPort{}, err
+	}
+
+	addrToUse, ok := netip.AddrFromSlice(upIPs[0])
+	if !ok {
+		return netip.AddrPort{}, err
+	}
+
+	upstream := netip.AddrPortFrom(addrToUse, 443)
+
+	return upstream, nil
+}
+
+// Serve will start serving connections caught by a Server
+func (s *Server) Serve() error {
+	var ok bool
+
+	defer func() {
+		if !ok {
+			for _, k := range s.ds {
+				_ = k.Close
+			}
+		}
+	}()
+
+	errHTTP := s.prepareRedir()
+	if errHTTP != nil {
+		return errHTTP
+	}
+
+	err := s.prepareDispatchers()
+	if err != nil {
+		return err
+	}
+	ok = true
+	return s.wg.Wait()
+}
+
+func (s *Server) prepareRedir() error {
+	if s.redir {
+		s.wg.Go(func() error {
+			return http.ListenAndServe(":80",
+				middleware.NewHTTPSRedirectHandler(0))
+		})
+	}
+	return nil
+}
+
+func (s *Server) prepareDispatchers() error {
+	for _, p := range s.ps {
+		dispatcher := &sni.Dispatcher{
+			GetHandler: getDefaultTLSHandler,
+		}
+		s.ds = append(s.ds, dispatcher)
+		ll, err := net.Listen("tcp", p)
+		if err != nil {
+			return err
+		}
+		s.wg.Go(func() error {
+			return dispatcher.Serve(ll)
+		})
+	}
+	return nil
+}
+
+// Err returns the error that caused the cancellation, if any
+func (s *Server) Err() error {
+	if err, ok := s.err.Load().(error); ok {
+		return err
+	}
+	return nil
+}
+
+// Wait waits until all workers are done
+func (s *Server) Wait() error {
+	s.wg.Wait()
+	return s.Err()
+}
+
+func (s *Server) tryCancel(err error) {
+	// once
+	if s.cancelled.CompareAndSwap(false, true) {
+		const msg = "Initiating shutdown"
+
+		if err != nil {
+			fmt.Printf("%s: %s", msg, err.Error())
```

Comment:
please use our logger

Reply:
I don't want a logger, tlsfwd is not a proxy, it is just a convenience one shot thing to test stuff